File: DataStructure/Utildata.js


class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.size = 0;
    }
    InsertFront(data) {
        var newNode = new Node(data);
        newNode.next = this.head;
        this.head = newNode;
    }
    InsertSorted(data) {
        var newNode = new Node(data);
        var current = this.head;
        var previous = this.head;
        if (this.head == null) {
            newNode.next = this.head;
            this.head = newNode;
            return true;
        }
        else {
            if (current.data >= newNode.data) {
                newNode.next = this.head;
                this.head = newNode;
                console.log(current.data);
                return true;
            }
            while (current.data < newNode.data) {
                if (current.next == null) {
                    break;
                }
                previous = current;
                current = current.next;
            }
            if (previous.data < newNode.data && newNode.data < current.data) {
                newNode.next = previous.next;
                previous.next = newNode;
                return true;
            }
            else {
                newNode.next = current.next;
                current.next = newNode;
                return true;
            }
        }
    }
    /***************************************insertNextToKey method************************/
    InsertNextTOKey(key, data) {
        var current = this.head;
        while (current.data != key) {
            if (current.next == null) {
                break;
            }
            current = current.next;
        }
        if (current.data == key) {
            var newNode = new Node(data);
            newNode.next = current.next;
            current.next = newNode;
            console.log("element is added to list");
            return true;
        }
        else {
            console.log("key is not in the list");
            return false;
        }
    }
    /*************************************DisplayStringList method************************/
    DisplayIntList() {
        var current = this.head;
        if (head != null) {
            while (current != null) {
                console.log(current.data + "-->");
                current = current.next;
            }
            console.log("null");
            console.log();
        }
        else {
            console.log("List is empty");
        }
    }
    /*************************************DisplayStringList method************************/
    DisplayStringList() {
        var current = this.head;
        // console.log(this.head);
        var arr = " ";
        while (current != null) {
            // var arr = " ";
            arr += current.data + "->";
            current = current.next;


        }
        console.log(arr);
        console.log();
    }

    /***************************************ReturnString method************************/
    ReturnString() {
        var stringText = "";
        var current = this.head;
        if (this.head != null) {
            stringText += current.data;
            current = current.next;
            while (current != null) {
                stringText = stringText + " " + current.data;
                current = current.next;
            }
            return stringText;
        }
        else {
            console.log("List is empty");
            return null;
        }
    }
    /***************************************ReturnString method************************/
    ReturnNumberString() {
        var stringText = "";
        var current = head;
        if (this.head != null) {
            stringText += current.data;
            current = current.next;
            while (current != null) {
                stringText = stringText + " " + current.data;
                current = current.next;
            }
            return stringText;
        }
        else {
            console.log("List is empty");
            return null;
        }
    }
    /***************************************ReturnString method***********************
    public int ReturnIntArray()
    {
        int[] intArray = new int[1000];
        Node current = head;
        if (head != null)
        {
            int i = 0;
            while (current != null)
            {
                intArray[i] = current.data;
                current = current.next;
                i++;
            }
            return intArray;
        }
        else
        {
            console.log("List is empty");
            return null;
        }
    }
    */
    /***************************************Search-string method************************/
    Search(key) {
        var current = this.head;
        if (this.head == null) {
            console.log("List is empty so none element deleted");
            return false;
        }
        while (current.data != key) {
            if (current.next == null) {
                break;
            }
            current = current.next;
        }
        if (current.data == key) {
            console.log("element found in list");
            return true;
        }
        else {
            console.log("element is not in list");
            return false;
        }
    }
    /***************************************Search-integer method************************/
    SearchInt(key) {
        var current = this.head;
        if (this.head == null) {
            console.log("List is empty so none element deleted");
            return false;
        }
        while (current.data != key) {
            if (current.next == null) {
                break;
            }
            current = current.next;
        }
        if (current.data == key) {
            console.log("element found in list");
            return true;
        }
        else {
            console.log("element is not in list");
            return false;
        }
    }
    /***************************************DeleteFront method************************/
    DeleteFront() {
        if (this.head == null) {
            console.log("List is empty so none element deleted");
        }
        else if (this.head.next == null) {
            this.head = null;
        }
        else {
            this.head.data = this.head.next.data;
            this.head.next = this.head.next.next;
        }
    }
    /***************************************DeleteKeyElement(string) method************************/
    DeleteKeyElement(key) {
        var current = this.head;
        var previous = this.head;
        if (current.data == key) {
            this.head.data = this.head.next.data;
            this.head.next = this.head.next.next;
            console.log("1st element deleted from the list");
            return true;
        }
        else {
            current = current.next;
        }
        while (current.data != key) {
            if (current.next == null) {
                break;
            }
            previous = current;
            current = current.next;
        }
        if (current.data == key) {
            previous.next = current.next;
            console.log("element  deleted from the list");
            return true;
        }
        else {
            console.log("element is not in list");
            return false;
        }
    }
    /***************************************DeleteKeyElement(string) method************************/
    DeleteIntKeyElement(key) {
        var current = this.head;
        var previous = this.head;
        if (current.data == key) {
            this.head.data = this.head.next.data;
            this.head.next = this.head.next.next;
            console.log("1st element deleted from the list");
            return true;
        }
        else {
            current = current.next;
        }
        while (current.data != key) {
            if (current.next == null) {
                break;
            }
            previous = current;
            current = current.next;
        }
        if (current.data == key) {
            previous.next = current.next;
            console.log("element  deleted from the list");
            return true;
        }
        else {
            console.log("element is not in list");
            return false;
        }
    }
    // end of all methods.
}





/**
 * 
 * @param {*} readlinessync : fetch the value fom user :
 * @param {*} module : Export module from server :
 * 
 */
var readlinessync = require('readline-sync')
module.exports = {
    LinkedList,

    //**
    //* 
    //* @param {*} str1 :
    //* @param {*} str2 :
    //*/

    isAnagram(str1, str2) {
        //var str1=13,13,11;
        //var str2=23,11,19;
        var check = false;
        var count = 0;
        var string1 = str1.toString().split("");
        string1.sort();
        var string2 = str2.toString().split("");
        string2.sort();

        var l1 = string1.length;
        //console.log(l1);
        if (string1.length != string2.length) {
            return false;
        }

        else {
            for (var i = 0; i < l1; i++) {
                if (string1[i] == string2[i]) {
                    count++;
                }
            }
            if (count == string1.length) {
                return true;
                //console.log("string are anagram");
            }
            else {
                return false;
                //console.log("string are not anagram");
            }
        }
    },
    sort(no) {
        var arr = no;
        for (var i = 0; i < arr.length - 1; i++) {
            for (var j = i + 1; i < arr.length; j++) {
                if (no[i] < no[j]) {
                    var temp = no[i];
                    no[i] = no[j];
                    no[j] = temp;
                }
            }
        }
        return (arr);
    },















}

File: datastructure.java/Utildata.java
package com.bridgelabz.fellowshipprograms.datastructure;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Scanner;

import com.bridgelabz.fellowshipprograms.datastructure.linkedlist.Node;


public class Utildata 
{
	//static scanner
	static Scanner sc = new Scanner(System.in); 
	static linkedlist list = new linkedlist();
	public String readWords() throws FileNotFoundException 
	{
		File f1 = new File("/home/admin1/Desktop/FellowshipProgramsDeepak/FellowshipPrograms/src/com/bridgelabz/fellowshipprograms/files/thakur.txt");
		String str = null;
		String name="";
		while (sc.hasNextLine())
		{
			str = sc.nextLine();
			//System.out.println(str);
			String words[] = str.split(" ");
			
			
			name=Arrays.toString(words);
			list.addNode(name);
			//System.out.println(name);
		}
		return name;
		
	}	
		
	/*
	 --------------------------------------------------------------------------------
	 Linked list operations on integer values
	 */	
	
	
		class Node
		{
			int datad;
			String data;
			Node next;
			int d;
			
			//constructor 
			Node(int d)
			{
				this.datad = d;
				this.next = null;
			}
			Node(String data)
			{
				this.data = data;
				this.next = null;
			}
		}
		public static Node head = null;
		public Node last = null;
		
		// add nodes to list	
	    
		public void addNode(int d)
		{
			//System.out.println(d);
			Node newnode = new Node(d);
			if(head == null)
			{
				head = newnode;
				last = newnode;
				newnode.datad = d;
				System.out.println(d);
			}
			else
			{
				last.next = newnode;
				last = newnode;
				newnode.datad = d;
				System.out.println(d);
			}
		}
	    /*
	     * print list nodes
	     */
		public void prinInttList()
		{
			Node current = head;
			if(head == null)
			{
				System.out.println("List is empty");
			}
			System.out.println("Nodes of list are :");
			System.out.println();
			while(current.next != null)
			{
			
				System.out.print(current.datad+" ");
				current = current.next;
			}
			System.out.println();
		}
		
		/*
		 * function to sort numbers which are take from file
		 */
		public void sortNum(int [] num)
		{
			int i=0;
			int j=0;
			for(i=0; i<=num.length; i++)
			{
				for(j=i+1; j<num.length; j++)
				{
					if(num[i] > num[j])
					{
						int tmp = num[i];
						num[i] = num[j];
						num[j] = tmp;
					}
				}
			}
			
			for(i=0; i<num.length; i++)
			{
				//System.out.print(num[i]+" ");
				addNode(num[i]);
			}
		}
		/*
		 * function to search given no in list
		 */
		public void searchInt(int no)
		{
			int i=1;
			int flag = 0;
			Node current = head;
			
			if(head == null)
			{
				System.out.println("List is empty");
			}
			while(current != null)
			{
				if(current.d == no)
				{
					flag = 1;
	     			//System.out.println("The element found at position "+i);
				}
				current = current.next;
				i++;
				if(flag == 1)
				{
					System.out.println("The element found at position "+i);
				}
				else
				{
					System.out.println("Element not found");
				}

			}
		}
		
		/*
		  * remove element from list
		  * by given key
		  */
		public void removeInt(int key)
		{
			 Node current = head;
			 Node prev = null;
	   	     if(head == null)
			 {
				System.out.println("List is empty");
			 }
			 else if(current.next != null && current.datad == key)
	   	     {
				current = current.next;
			 }
			 while(current.next != null && current.datad != key)
			 {
				prev= current.next;
				current = current.next;
			 }
				prev.next = current.next;
		}
			
	/*
	 * ----------------------------------------------------------------------------------------
	 *            Linked List for String Data
	 */
			
		///add nodes to list
		
		
			public void addStringNode(String data)
			{
				Node newnode = new Node(data);
				if(head == null)
				{
					head = newnode;
					last = newnode;
				}
				else
				{
					last.next = newnode;
					last = newnode;
				}
			}	

	///print list nodes
			
			public void printList() 
			{
				Node current = head;
				if(head == null)
				{
					System.out.println("List is empty");
				}
				System.out.println("Nodes of list are :");
				while(current != null)
				{
					System.out.print(current.data+" ");
					current = current.next;
				}
				
			}
		
			
			
		///function to count size of list
			 
			public void listSize()
			{
				int count = 0;
				Node current = head;
				while(current != null)
				{
					count++;
					current = current.next;
				}
				System.out.println("Size of list is : "+count);
			}
		
	
		///method to search the word from list
			
			
			 public boolean searchString(String key)
			 {
				int i=1;
				int flag = 0;
				Node current = head;
				
				if(head == null)
				{
//					System.out.println("List is empty");
					return false;
				}
				while(current != null)
				{
					if(current.data == key)
					{
						return true;
		     			//System.out.println("The element found at position "+i);
					}
					current = current.next;
					i++;
//					if(flag == 1)
//					{
//						System.out.println("The element found at position "+i);
//					}
//					else
//					{
//						System.out.println("Element not found");
//					}

				}
				return false;
			 }
			/*
			  * remove element from list
			  * @param key
			  */
			static void removeString(String key)
			{
				 Node current = head;
				 Node prev = null;
		    	 if(head == null)
				 {
					System.out.println("List is empty");
				 }
				 else if(current != null && current.data == key)
		    	 {
					current = current.next;
				 }
				 while(current != null && current.data != key)
				 {
					prev= current;
					current = current.next;
				 }
					prev.next = current.next;
			}
				
			
	/*
	 * ------------------------------------------------------------------------
	 *               Stack using linked list		
	 */
			//class to create stack
			
			StackNode top;
			class StackNode
			{
				int data;
				StackNode next;
				StackNode top; 
				//constructor to set top to null
				StackNode()
				{
					this.top = null;
				}
			}
			
			/*
			 function to add elements in stack
			 add data x into stack
			 */
			public void push(int x)
			{
				StackNode tmp = new StackNode();
				
				tmp.data = x;
				tmp.next = top;
				top = tmp;
			}
			
		/*
		function to remove top elements from stack and reutun stack data
		 */
			public int pop()
			{
				if(top == null)
				{
					System.out.println("Stack is empty");
				}
				int tmp =top.data;
				top = top.next;
				System.out.println("Top element "+tmp+" is removed from stack");
				return tmp;
				
			}
			
		/*
		 function print top element of stack
		*/
			public int peak()
			{
				if(top == null)
				{
					System.out.println("Stack is empty");
				}
				System.out.println(top.data);
				return top.data;
				
			}
			
		/*
		function to show all stack elements
		*/
			public void printStack()
			{
				if(top == null)
				{
					System.out.println("Stack is empty");
				}
				StackNode tmp = top;
				System.out.println("Stack elements are : ");
				while(tmp != null)
				{
					System.out.println(tmp.data+" ");
					tmp = tmp.next;
				}
			}
		/*
		 function to count size of stack
		 */
			public void countSize()
			{
				int count = 0;
				StackNode temp = top;
				while(temp != null)
				{
					count++;
					temp = temp.next;
				}
				System.out.println("Size of stack is : "+count);
			}
			
	/*
	function to check stack is empty or not return null
	*/
			public boolean isEmpty()
			{
				return top==null;	
			}
			
		/*
		function to push char into stack
		*/
			public void pushChar(char ch[])
			{
				//int n = ch;
				int i=0;
				//char ch[] = expr.toCharArray();
				StackNode tmp = new StackNode();
				
				tmp.data = ch[i];
				tmp.next = top;
				top = tmp;
				i++;
			}
			
			/**
			 function to remove top char from stack and return eliminated data
			
			 */
			public char popChar()
			{
				if(top == null)
				{
					System.out.println("Stack is empty");
				}
				char tmp =(char) top.data;
				top = top.next;
				System.out.println("Top element "+tmp+" is removed from stack");
				return tmp;
						
			}
	  /*
	  function to check whether expression is balanced or not		
	  */
			public void checkBalancedExpr(String expr)
			{
				int n = expr.length();
				for(int i=0; i<n; i++)
				{
					//check opening parenthesis to push in stack 
					if(expr.charAt(i)== '{' || expr.charAt(i) == '[' || expr.charAt(i) == '(')
					{
						push(expr.charAt(i));
					}
					//to check current char and top of stack 
					//if matches pair 
					//remove from stack
					else if(top == null && (
							(expr.charAt(i)==']' && peak()=='[' )
							|| (expr.charAt(i)==')' && peak()=='(' )
							|| (expr.charAt(i)=='{' && peak()=='}')))
					{
						popChar();
					}
					else
					{
						push(expr.charAt(i));
					}
				}
				if(isEmpty() == true)
				{
					System.out.println("Expression is balanced");
				}
				else
				{
					System.out.println("Expression is not balanced");
				}
			}
	
	
	//function to take string from user
			public String readString()
			{
				// TODO Auto-generated method stub
				return sc.nextLine();
			}
			
			
			//function to take input from user
			public int readInt() 
			{
				// TODO Auto-generated method stub
				return sc.nextInt();
			}

	
	

}

